pkg/proxy/handler.go

```diff
@@ -158,5 +145,50 @@ func newReverseProxy(host string) *httputil.ReverseProxy {
 			AdditionalProxyRewrite(r)
 		},
 	}
-	return proxy
+
+	proxy.ModifyResponse = func(r *http.Response) error {
+		// Record the response for metrics.
+		pr.status = r.StatusCode
+
+		// This point is reached if a response code is received.
+		if h.isRetryCode(r.StatusCode) && pr.attempt < h.MaxRetries {
+			// Returning an error will trigger the ErrorHandler.
+			return ErrRetry
+		}
+
+		return nil
+	}
+
+	proxy.ErrorHandler = func(w http.ResponseWriter, r *http.Request, err error) {
+		// This point could be reached if a bad response code was sent by the backend
+		// or
+		// if there was an issue with the connection and no response was ever received.
+		if err != nil && pr.attempt < h.MaxRetries {
```

Comment:
It would be good to abort early when the request context is canceled. I have debugged retries in #65 that never hit a backend
```suggestion
		if err != nil &&
			r.Context().Err() == nil &&
			pr.attempt < h.MaxRetries {
```

Reply:
It looks like there are checks for the context being done in the reverse proxy .ServeHTTP function. Do you know what cases it would fix to also do the check here?